src/terms.jl

```diff
@@ -0,0 +1,524 @@
+part of the DSL, it replaces the expression with a call to `capture_call` with 
+arguments: 
+
+* `f_orig`: the original function that was called.
+* `f_anon`: an anonymous function that calls the original expression, replacing 
+  each symbol with one argument to the anonymous function
+* `names`: the symbols from the original expression corresponding to each 
+  argument of `f_anon`
+* `ex_orig`: the original (quoted) expression before [`capture_call_ex!`](@ref).
+* `args_parsed`: a vector of the original arguments, wrapped in terms with the 
+  special formula DSL rules applied, as if `f_orig` was a special DSL call.
+
+The default behavior is to pass these arguments to the `FunctionTerm` constructor.
+This default behavior can be overridden by dispatching on `call`.  That is, to 
+change how calls to `myfun` in a formula are handled in the context of `MyModel`, 
+add a method:
+    
+    apply_schema(t::FunctionTerm{typeof(myfun)}, schema, ::Type{MyModel})
+
+If you simply want to pass the arguments to the original call, parsed as if the 
+call was a special, then you can use
+
+    apply_schema(t::FunctionTerm{typeof(myfun)}, schema, ::Type{MyModel}) =
+        apply_schema(myfun(t.args_parsed...), schema, MyModel)
+
+For this to work, `myfun(args::AbstractTerm...)` should return another 
+`AbstractTerm`, which has an `apply_schema` method already defined for it.
+"""
+capture_call(args...) = FunctionTerm(args...)
+
+extract_symbols(x) = Symbol[]
+extract_symbols(x::Symbol) = [x]
+extract_symbols(ex::Expr) =
+    is_call(ex) ? mapreduce(extract_symbols, union, ex.args[2:end]) : Symbol[]
+
+################################################################################
+# operators on Terms that create new terms:
+
+
+Base.:~(lhs::TermOrTerms, rhs::TermOrTerms) = FormulaTerm(lhs, rhs)
+
+Base.:&(terms::AbstractTerm...) = InteractionTerm(terms)
+Base.:&(term::AbstractTerm) = term
+Base.:&(it::InteractionTerm, terms::AbstractTerm...) = InteractionTerm((it.terms..., terms...))
+
+Base.:+(terms::AbstractTerm...) = (unique(terms)..., )
+Base.:+(as::TupleTerm, b::AbstractTerm) = (as..., b)
+Base.:+(a::AbstractTerm, bs::TupleTerm) = (a, bs...)
+
+################################################################################
+# evaluating terms with data to generate model matrix entries
+
+"""
+    model_cols(t::AbstractTerm, data)
+
+Create a numerical "model columns" representation of data based on an
+`AbstractTerm`.  `data` can either be a whole table or a single row (in the form
+of a `NamedTuple` of scalar values).  Tables will be converted to a `NamedTuple`
+of `Vectors` (e.g., a `Tables.ColumnTable`).
+"""
+function model_cols(t, d::D) where D
+    Tables.istable(d) || throw(ArgumentError("Data of type $D is not a table!"))
+    model_cols(t, columntable(d))
+end
+
+"""
+    model_cols(ts::NTuple{N, AbstractTerm}, data) where N
+
+When a tuple of terms is provided, `model_cols` broadcasts over the individual 
```

Comment:
You mean using `model_cols.(ts, Ref(data))`?  It's a good point; the downside is that it would require doing a "defensive broadcast" everywhere a tuple of terms might appear (e.g., the RHS of a formula).  Which wouldn't be such a big deal I guess, as long as we also define `broadcastable(t::AbstractTerm) = t`.

Reply:
(This too is a holdover from an earlier implementation: before `MatrixTerm`, a tuple of terms' columns were all concatenated into a single matrix)

The "automatic broadcasting over tuples" occurs in `apply_schema`, too

Reply:
Do you think it will be common for users to write code that can pass either a single term or a tuple of terms? I would think most of the time you expect a specific type for a given model family.

Reply:
I think this issue is more likely to come up internally, in e.g. `apply_schema`, which is applied recursively.  So if it's applied to the RHS of a formula, you might get one term or many.

Reply:
OK. But you know how to handle it using broadcast, so that's not a problem. :-)